### ISSUE TITLE
Suggestion: Expose a parameter of  `ShouldMatch` for `<NavLink />` 

### ISSUE THREAD

**Issue by newbienewbie** [aspnet/AspNetCore#18163]
### Is your feature request related to a problem? Please describe.

I want to custom the matching behavior of `<NavLink />`.  

Let's say we want to process a url with a hash(See [this thread on SO](https://stackoverflow.com/q/59614216/10091607)) : 
```
http://localhost/#an-element-id
```
Sometimes we might think this url matches `http://localhost` , while some other users don't think so.

It would be great that we are allowed to custom this matching behavior. Usually, we can change the behavior by inheritance or composition.  However, the [matching logic](https://github.com/aspnet/AspNetCore/blob/c17ce436b8703470231e7979cead042f5682c3f5/src/Components/Web/src/Routing/NavLink.cs#L110-L129) is hardcoded for the present.

1.  There's no `public`/`protected` `ShouldMatch(string)` method exposed. So it's difficult to achieve it by inheritance
2. There's no `ShouldMatch` parameter exposed. It would be great if we can pass a `ShouldMatch` `[Parameter]` to this component.

### Describe the solution you'd like

Expose a `[Parameter]` of `ShouldMatch(string)`, if user provides such a parameter, use this one. Otherwise, use the default logic.

**Comment by mkArtakMSFT** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by TanayParikh** [comment]
Just confirming my understanding of your proposal; this `ShouldMatch` would be a:

```csharp
        /// <summary>
        /// Gets or sets an action facilitating custom matching behavior.
        /// </summary>
        [Parameter]
        public Action<string, bool> ShouldMatch { get; set; }
```

wouldn't this conflict with the `NavLinkMatch` parameter?

https://github.com/dotnet/aspnetcore/blob/c17ce436b8703470231e7979cead042f5682c3f5/src/Components/Web/src/Routing/NavLink.cs#L48-L52

cc/ @javiercn

**Comment by DNFL** [comment]
I agree with this proposal, and I also want to extend the built-in url matching logic.

Here's a scene:
Because all url changes in Blazor are rewritten as navigation, it is not feasible to locate the element of the specified id by the hash value, which may be due to the need for hash routing.
So, I use navigation change events to listen for hash values and scroll to the target element through .Net calls Js to achieve the hash location function of traditional web applications.
However, `NavLink` match logic, `Match.All` and `Match.Prefix` cannot adapt to the existence of hash values. The former treats hash values as part of the path, while I need to ignore hash values, which can lead to confusion such as '/' and '/index''.

So, I want to extend the match logic to support `Match.AllWithoutHash`.
At the same time, `Match.AllByHash` is also possible in cases where hash routing may be supported in the future.

ps：sorry for my poor English, this is translated by may not very smart MT(machine translater)

**Comment by ilonatommy** [comment]
Closing as resolved.

The core ask here was to make `NavLink`s URL matching customizable instead of hardcoded, via either (1) an overridable `ShouldMatch`  method or (2) a `ShouldMatch` parameter.

`NavLink.ShouldMatch` is now a `protected virtual`  method (#59903, refined by #61151), so the matching logic is fully customizable through inheritance. The same work also made the default `NavLinkMatch.All`  ignore the query string and fragment by default (a frequent related complaint), with an `AppContext`  switch to restore the old behavior if needed.

Customize matching by overriding `ShouldMatch` (this covers hash routing, query/fragment-agnostic, regex, etc.):

```
 public class HashAgnosticNavLink : NavLink
 {
     protected override bool ShouldMatch(string uriAbsolute)
         => uriAbsolute.Split('#')[0].Equals(MyTargetUri, StringComparison.OrdinalIgnoreCase);
 }
```

 If you'd prefer the parameter-based ergonomics originally suggested (option 2), you can add it in a few lines, with full add/narrow/replace control:

```
 public class MatchableNavLink : NavLink
 {
     [Parameter] public Func<string, bool>? MatchPredicate { get; set; }

     protected override bool ShouldMatch(string uriAbsolute)
         => MatchPredicate?.Invoke(uriAbsolute) ?? base.ShouldMatch(uriAbsolute);
 }

 <MatchableNavLink href="/dashboard"
                   MatchPredicate="@(uri => uri.Split('#')[0] == MyUri)">
     Dashboard
 </MatchableNavLink>
```

 Since this is fully achievable through the supported `ShouldMatch` override, we're keeping the public API minimal rather than adding a built-in `[Parameter]` . If you hit a scenario the override genuinely can't express, please open a new issue.
